python/docs/source/user_guide/python_packaging.rst

```diff
@@ -0,0 +1,201 @@
+
+
+There are multiple ways to ship the dependencies to the cluster:
+
+- Using PySpark Native Features
+- Using Zipped Virtual Environment
+- Using PEX
+
+
+Using PySpark Native Features
+-----------------------------
+
+PySpark allows to upload Python files (``.py``), zipped Python packages (``.zip``), and Egg files (``.egg``)
+to the executors by setting the configuration setting ``spark.submit.pyFiles`` or by directly
+calling :meth:`pyspark.SparkContext.addPyFile`.
+
+This is an easy way to ship additional custom Python code to the cluster. You can just add individual files or zip whole
+packages and upload them. Using :meth:`pyspark.SparkContext.addPyFile` allows to upload code
+even after having started your job.
+
+Note that it doesn't allow to add packages built as `Wheels <https://www.python.org/dev/peps/pep-0427/>`_ and therefore doesn't
+allowing to include dependencies with native code.
+
+
+Using Zipped Virtual Environment
+--------------------------------
+
+The idea of zipped environments is to zip your whole `virtual environment <https://docs.python.org/3/tutorial/venv.html>`_, 
+ship it to the cluster, unzip it remotly and target the Python interpreter from inside this zipped environment. Note that this
+is currently supported *only for YARN*.
+
+Zip Virtual Environment
+~~~~~~~~~~~~~~~~~~~~~~~
+
+You can zip the virtual environment on your own or use tools for doing this:
+
+* `conda-pack <https://conda.github.io/conda-pack/spark.html>`_ for conda environments
+* `venv-pack <https://jcristharif.com/venv-pack/spark.html>`_ for virtual environments
+
+Example with `conda-pack`:
+
+.. code-block:: bash
+
+    conda create -y -n conda_env -c conda-forge \
+      pyspark==3.0.1 pyarrow==0.15.1 pandas==0.25.3 conda-pack==0.4.0
+    conda activate conda_env
+    conda pack -f -o conda_env.tar.gz
+
+Upload to Spark Executors
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The unzip will be done by Spark when using target ``--archives`` option in spark-submit
+or setting ``spark.yarn.dist.archives`` configuration.
+
+Example with ``spark-submit``:
+
+.. code-block:: bash
+
+    export PYSPARK_DRIVER_PYTHON=python
+    export PYSPARK_PYTHON=./environment/bin/python
+    spark-submit --master=yarn --deploy-mode client \
+      --archives conda_env.tar.gz#environment app.py
+
+Example using ``SparkSession.builder``:
+
+.. code-block:: python
+
+    import os
+    from pyspark.sql import SparkSession
+    from app import main
+
+    os.environ['PYSPARK_PYTHON'] = "./environment/bin/python"
+    builder = SparkSession.builder.master("yarn").config(
+        "spark.yarn.dist.archives", "conda_env.tar.gz#environment")
+    spark = builder.getOrCreate()
+    main(spark)
+
+
+Using PEX
+---------
+
+`PEX <https://github.com/pantsbuild/pex>`_ is a library for generating ``.pex`` (Python EXecutable) files.
+A PEX file is a self contained executable Python environment. It can be seen as the Python equivalent of Java uber-JARs (a.k.a. fat JARs).
```

Comment:
```suggestion
A PEX file is a self-contained executable Python environment. It can be seen as the Python equivalent of Java uber-JARs (a.k.a. fat JARs).
```